### PR TITLE
Click on labels focus corresponding inputs

### PR DESCRIPTION
**Pull request by babgyy**
Currently, when clicking a `label`, nothing happens. Ideally, clicking a `label` should focus (or activate for checkboxes) the `input`.
This is done by having the `for` attribute of the `label` matching the `id` attribute of the `input`.
Rails does that with an `all_namespace_snake_cased` convention, which is implemented here.

This commit does that, and updates the sample output in the README.

Potential issues / incompatibilities with client configuration if commit is merged after upgrade :
- `ids` of `inputs` are now overriden by the `fields` helpers. This could cause issues if a customer expects a certain `id` on a given `input`. It might be safer to ignore this new wiring when a custom `id` is given in `props`. I can do the change if needed.
- `labels` `for` will change.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import React from "react"
 import PropTypes from "prop-types"
 
-export const nameWithContext = (Lower, prop = "name") => {
+export const nameWithContext = (Lower, prop = "name", isInput = true) => {
```
Comment (danott) [inline src/utils.js]: I'm noticing that `prop` and `isInput` are kind-of traveling as a pair? All of the calls pass one of these two argument pairs:

```
["name", true]
["htmlFor", false]
```

Would it make sense to keep the existing function arity, and check `if (prop === "htmlFor")`?

File: src/utils.js [hunk 2]
```diff
@@ -10,12 +10,24 @@ export const nameWithContext = (Lower, prop = "name") => {
       .map((field, index) => (index === 0 ? field : `[${field}]`))
       .join("")
 
+  const buildInputId = (namespaces, name = "") =>
```
Comment (danott) [inline src/utils.js]: As you mentioned, overriding the `id` prop could break existing usage.  Perhaps we can provide some affordance for this. My idea would be something like...

```jsx
const buildInputId = (namespaces, name = "") => {
  if (this.props.id) return this.props.id
  return [...namespaces, name].join("_")
}
```

Would that make sense?

File: src/utils.js [hunk 3]
```diff
@@ -10,12 +10,24 @@ export const nameWithContext = (Lower, prop = "name") => {
...
+    [...namespaces, name].join("_")
+
   const higher = (props, context) => {
     const replacedProp = buildInputName(
       context.railsFormNamespaces,
       props[prop]
     )
-    const replacedProps = Object.assign({}, props, { [prop]: replacedProp })
+    const inputId = buildInputId(context.railsFormNamespaces, props[prop])
```
Comment (danott) [inline src/utils.js]: It looks like we're not doing anything with this local variable? Do we want to be using this on lines 25 and 28 in place of calls to the same function?

**Review by danott — COMMENTED** [review]
This is a great change. Labels should definitely be wired up to affect their corresponding input. (This will actually fix a bug in one of the forms on my app! 😬)

I had a few questions that I'd love to hear your thoughts on? Thanks for being proactive on this!

File: src/utils.js (hunk 1 above)
Comment (babgyy) [reply src/utils.js]: Indeed, 
I was not sure about breaking the agnostic nature of the method `nameWithContext`, but I guess the `isInput` parameter breaks it just as same as checking `if (prop === "htmlFor")` and introduces confusion ^^

Your suggestion makes more sense.

File: src/utils.js (hunk 2 above)
Comment (babgyy) [reply src/utils.js]: Sounds good !

File: src/utils.js (hunk 3 above)
Comment (babgyy) [reply src/utils.js]: Oops my bad ! sorry 😬

**Comment by babgyy** [comment]
Implementing these improvements, I realized two things : 
- the `buildInputId` method doesn't have access to props (as is)
- returning `if(props.id)` is not always what we want. If a `label` has an `id`, we still want to build the actual `inputId` and set it as `htmlFor` props

I suggest we build the additionalProps anyway, but instead of merging the additionalProps into the original props like this 
```
const replacedProps = Object.assign({}, props, additionalProps)
```
we do the opposite :
```
const replacedProps = Object.assign({}, additionalProps, props)
```
so that the original props always take precedence.
What do you think ?

**Comment by babgyy** [comment]
I pushed a new commit with your suggestions !

File: src/utils.js (hunk 3 above)
Comment (danott) [reply src/utils.js]: No worries!

**Review by danott — APPROVED** [review]
This is excellent. Thank you for taking care of this!

**Comment by danott** [comment]
Released as [0.1.0-beta.16](https://www.npmjs.com/package/react-rails-form-helpers/v/0.1.0-beta.16)

**Pull request merged**